frostsnap_core/src/coordinator.rs

```diff
@@ -422,246 +423,314 @@ impl FrostCoordinator {
+                    Entry::Occupied(mut entry) => {
+                        match entry.get_mut() {
+                            KeyGenState::WaitingForResponses {
+                                input_aggregator,
                                 device_to_share_index,
-                                pending_key_name: pending_key_name.to_string(),
-                                purpose: *purpose,
-                                certifier,
-                                coordinator_keypair: *coordinator_keypair,
-                            };
-
-                            self.pending_keygens.insert(keygen_id, new_state);
+                                ..
+                            } => {
+                                let device_to_share_index_cloned = device_to_share_index.clone();
+                                let cert_scheme =
+                                    certpedpop::vrf_cert::VrfCertScheme::<Sha256>::new(
+                                        crate::message::keygen::VRF_CERT_SCHEME_ID,
+                                    );
+                                let share_index = device_to_share_index_cloned.get(&from).ok_or(
+                                    Error::coordinator_invalid_message(
+                                        message_kind,
+                                        "got share from device that was not part of keygen",
+                                    ),
+                                )?;
+
+                                input_aggregator
+                                    .add_input(
+                                        &schnorr_fun::new_with_deterministic_nonces::<Sha256>(),
+                                        // we use the share index as the input generator index. The input
+                                        // generator at index 0 is the coordinator itself.
+                                        (*share_index).into(),
+                                        *response.input,
+                                    )
+                                    .map_err(|e| {
+                                        Error::coordinator_invalid_message(message_kind, e)
+                                    })?;
+
+                                let mut outgoing = vec![CoordinatorSend::ToUser(
+                                    CoordinatorToUserMessage::KeyGen {
+                                        keygen_id,
+                                        inner: CoordinatorToUserKeyGenMessage::ReceivedShares {
+                                            from,
+                                        },
+                                    },
+                                )];
+
+                                if input_aggregator.is_finished() {
+                                    // Remove the entry to take ownership
+                                    let state = entry.remove();
+                                    if let KeyGenState::WaitingForResponses {
```

Comment:
Inspecting the entry style does force this double match-pattern, fine though

Reply:
yes I couldn't stand it either -- I added 940def0cbcf469a8ecc8c5a69e2b9bd8ed236492 to clean it up.